### PR TITLE
Update to cardano-node 1.35-rc2

### PR DESCRIPTION
**Pull request by Anviking**
### Overview

- Updates dependencies to target `cardano-node 1.35-rc2`
- Adds support to make the wallet _mostly_ work in Babbage
- Rework of nix setup: we now derive nix files from `cabal.project` rather than the weird `stack.yaml`-based approach we had before.

### Detailed todo-list

- [x] Get nix happy with the new dependencies
- [x] Make `cabal build all` work with modulo warnings
- [x] Update to more recent revisions
    - [x] Fix compilation in tests again
- [x] Make `cabal build all` work with _no_ warnings
    - [x] Fix puzzling persistent TH name shadowing warnings -> fixed by sam?
- [x] Ensure existing tests pass
    - [x] unit
    - [x] Look closer at balanceTx golden unit failure after first seeing if balanceTx integration tests pass on both alonzo and Babbage
    - [x] Integration (on alonzo like on master)
        - [x] Was blocked by readTVar, fixed with https://github.com/input-output-hk/io-sim/pull/6#pullrequestreview-994757891
        - [x] Fix failures relating to Babbage era -> fixed with a hack https://github.com/input-output-hk/cardano-wallet/pull/3280/commits/9e2568e8058e2e4b995363805f4381ce9976108c
        - [x] Fix cardano-wallet-core Aeson failures
- [x] More closely review code-changes
- [x] Clean up git commit history
    - [x] Flatten merge commits
    - [x] Squash back-and-forth nix and revision changes
    - [x] Rebase off master 
- [x] Ensure revisions in stack.yaml and cabal.project are updated consistently 
    - Alternatively remove `stack.yaml` and ensure ./nix/regenerate.sh still works and that we know how to update sha256map.nix — #3328 but without the sha256map.nix
- [x] Ensure CI is completely green: [Hydra jobset](https://hydra.iohk.io/jobset/Cardano/cardano-wallet-pr-3280), ADP-1888
    - There may be CI specific problems with nix, windows and more (there probably are)
    - [x] fix hydra wine TH error e.g. `$(TH.lift =<< TH.runIO (readFile "app/foo.txt"))`
          - seems relevant https://github.com/input-output-hk/haskell.nix/pull/1508
          - seems gone now
    - [x] `Ambiguous module name` https://hydra.iohk.io/build/15469467
        - https://input-output-rnd.slack.com/archives/C48HA39RS/p1654388106509579
        - Bumping Haskell.nix again should help?
        - Didn't seem like it worked...
        - https://input-output-rnd.slack.com/archives/CAP8NM7N0/p1654828014483969
     - [x] Remove-iserv problem https://hydra.iohk.io/build/15604435 
         - Update from Hamish https://input-output-rnd.slack.com/archives/CAP8NM7N0/p1655213059547499?thread_ts=1654828014.483969&cid=CAP8NM7N0
     - [x] Docs shell - we don't need explicit emanote input in nix config
     - [x] Profiled shell:  `recursionzmschemeszm5zi2zi2zi2zm6NC6xZZx2tI3826DrVhqI9v_DataziFunctorziFoldableziTH_zdfMakeBaseFunctorNamezuzdcmakeBaseFunctor_closure` https://hydra.iohk.io/build/15694577/nixlog/2/tail
     - [x] musl cli goldens https://hydra.iohk.io/build/15694618/nixlog/82
     - [x] `hie-bios: hie-bios: this operation requires -fexternal-interpreter
     - [x] `Ambiguous module name 'Options.Applicative':` in HLS - now we at least have an ugly `-XPackageImport` workaround.
- [x] Update node revision in readme.md

### Things we need eventually, but not strictly for this PR

> -  Update to final 1.35 tag once it is released - can be done in separate PR
> -  Run existing integration tests on both Alonzo and Babbage eras in CI
> - Depend on cardano-addresses with aeson2 support merged, tagged and released
> - Depend on a cardano-sl-x509 where aeson 2 support is merged to master
> - Consider appropriateness Tx deserialisation hack https://github.com/input-output-hk/cardano-wallet/pull/3280/commits/9e2568e8058e2e4b995363805f4381ce9976108c

Moved to jira tasks in https://input-output.atlassian.net/browse/ADP-1538

### Comments

Includes #3258, #3293

### Issue Number

ADP-1744

### CODE REVIEW

File: lib/core/test/unit/Cardano/Api/GenSpec.hs [hunk 1]
```diff
@@ -1184,6 +1195,28 @@ genTxOutDatumCoverage era datum =
...
 
+genTxOutReferenceScriptCoverage :: CardanoEra era -> ReferenceScript era -> Property
+genTxOutReferenceScriptCoverage era refScript =
+    case refInsScriptsAndInlineDatsSupportedInEra era of
+        Nothing ->
+            (refScript == ReferenceScriptNone)
+            & label "reference scripts not generated in unsupported era"
+            & counterexample ( "reference scripts were generated in unsupported "
+                               <> show era
+                             )
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Api/GenSpec.hs]:
```suggestion
            & counterexample $ mconcat
                [ "reference scripts were generated in unsupported "
                , show era
                ]
```

File: lib/core/src/Network/Wai/Middleware/Logging.hs [hunk 2]
```diff
@@ -277,7 +279,7 @@ instance ToJSON HandlerLog where
 sanitize :: [Text] -> ByteString -> Text
 sanitize keys bytes = encode' $ case decode' bytes of
     Just (Object o) ->
-        Object (foldr (HM.adjust obfuscate) o keys)
+        Object (Aeson.fromHashMap (foldr (HM.adjust obfuscate) (Aeson.toHashMap o) (map Aeson.fromText keys)))
```
Comment (jonathanknowles) [inline lib/core/src/Network/Wai/Middleware/Logging.hs]:
```suggestion
        Object
            $ Aeson.fromHashMap
            $ foldr
                (HM.adjust obfuscate)
                (Aeson.toHashMap o)
                (map Aeson.fromText keys)
```

File: lib/core/test/unit/Cardano/Api/GenSpec.hs [hunk 3]
```diff
@@ -379,6 +381,18 @@ genTxIxCoverage (TxIx ix) = unsignedCoverage (maxBound @Word16) "txIx" ix
...
     arbitrary = genTxIndex
 
+genTxIxCoverage' :: Ledger.TxIx -> Property
+genTxIxCoverage' (Ledger.TxIx ix) = unsignedCoverage (maxBound @Word16) "txIx" ix
+
+instance Arbitrary Ledger.TxIx where
+    arbitrary = genTxIx
+
+genCertIxCoverage :: Ledger.CertIx -> Property
+genCertIxCoverage (Ledger.CertIx ix) = unsignedCoverage (maxBound @Word16) "certIx" ix
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Api/GenSpec.hs]:
```suggestion
genCertIxCoverage (Ledger.CertIx ix) =
    unsignedCoverage (maxBound @Word16) "certIx" ix
```

File: lib/core/test/unit/Cardano/Api/GenSpec.hs [hunk 4]
```diff
@@ -379,6 +381,18 @@ genTxIxCoverage (TxIx ix) = unsignedCoverage (maxBound @Word16) "txIx" ix
 instance Arbitrary TxIx where
     arbitrary = genTxIndex
 
+genTxIxCoverage' :: Ledger.TxIx -> Property
+genTxIxCoverage' (Ledger.TxIx ix) = unsignedCoverage (maxBound @Word16) "txIx" ix
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Api/GenSpec.hs]:
```suggestion
genTxIxCoverage' (Ledger.TxIx ix) =
    unsignedCoverage (maxBound @Word16) "txIx" ix
```

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 5]
```diff
@@ -1370,7 +1371,7 @@ instance FromJSON SchemaApiErrorCode where
     parseJSON = withObject "SchemaApiErrorCode" $ \o -> do
         vals <- forM (fmap showConstr $ dataTypeConstrs $ dataTypeOf NoSuchWallet)
             $ \n -> do
-                (r :: Maybe Yaml.Value) <- o .:? T.pack (toSchemaName n)
+                (r :: Maybe Yaml.Value) <- o .:? Aeson.fromString (toSchemaName n)
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]:
```suggestion
                (r :: Maybe Yaml.Value) <-
                    o .:? Aeson.fromString (toSchemaName n)
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 6]
```diff
@@ -19,11 +19,16 @@
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
 {-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs]:
```suggestion
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 7]
```diff
@@ -1067,8 +1084,10 @@ toAlonzoPParams
                 (Ledger.boundRational protocolParamTreasuryCut)
         , Alonzo._coinsPerUTxOWord =
             toShelleyLovelace utxoCostPerWord
-        , Alonzo._costmdls =
-            toAlonzoCostModels protocolParamCostModels
+        , Alonzo._costmdls = either
+                                  (\e -> error $ "toAlonzoPParams: invalid cost models, error: " <> e)
+                                  id
+                                  (toAlonzoCostModels protocolParamCostModels)
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs]:
```suggestion
        , Alonzo._costmdls = either
            (\e -> error $ "toAlonzoPParams: invalid cost models, error: " <> e)
            id
            (toAlonzoCostModels protocolParamCostModels)
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 8]
```diff
@@ -1090,21 +1109,24 @@ toAlonzoPParams
...
-            Alonzo.PlutusV2
+      -> Either String Alonzo.CostModels
+    toAlonzoCostModels m = do
+      f <- mapM conv $ Map.toList m
+      Right . Alonzo.CostModels $ Map.fromList f
+     where
+      conv :: (Cardano.AnyPlutusScriptVersion, Cardano.CostModel) -> Either String (Alonzo.Language, Alonzo.CostModel)
+      conv (anySVer, cModel )= do
+        alonzoCostModel <- toAlonzoCostModel cModel (toAlonzoScriptLanguage anySVer)
+        Right (toAlonzoScriptLanguage anySVer, alonzoCostModel)
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs]:
```suggestion
    toAlonzoCostModels m = do
        f <- mapM conv $ Map.toList m
        Right . Alonzo.CostModels $ Map.fromList f
      where
        conv
            :: (Cardano.AnyPlutusScriptVersion, Cardano.CostModel)
            -> Either String (Alonzo.Language, Alonzo.CostModel)
        conv (anySVer, cModel) = do
            alonzoCostModel <-
                toAlonzoCostModel cModel (toAlonzoScriptLanguage anySVer)
            Right (toAlonzoScriptLanguage anySVer, alonzoCostModel)
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 9]
```diff
@@ -1090,21 +1109,24 @@ toAlonzoPParams
...
+
+    toAlonzoCostModel :: Cardano.CostModel -> Alonzo.Language -> Either String Alonzo.CostModel
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs]:
```suggestion
    toAlonzoCostModel
        :: Cardano.CostModel
        -> Alonzo.Language
        -> Either String Alonzo.CostModel
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 10]
```diff
@@ -1090,21 +1109,24 @@ toAlonzoPParams
...
+    toAlonzoCostModel (Cardano.CostModel m) l = Alonzo.mkCostModel l m
+
+    toAlonzoScriptLanguage :: Cardano.AnyPlutusScriptVersion -> Alonzo.Language
+    toAlonzoScriptLanguage (Cardano.AnyPlutusScriptVersion Cardano.PlutusScriptV1) =
+        Alonzo.PlutusV1
+    toAlonzoScriptLanguage (Cardano.AnyPlutusScriptVersion Cardano.PlutusScriptV2) =
+        Alonzo.PlutusV2
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs]:
```suggestion
    toAlonzoScriptLanguage :: Cardano.AnyPlutusScriptVersion -> Alonzo.Language
    toAlonzoScriptLanguage
        (Cardano.AnyPlutusScriptVersion Cardano.PlutusScriptV1) =
            Alonzo.PlutusV1
    toAlonzoScriptLanguage
        (Cardano.AnyPlutusScriptVersion Cardano.PlutusScriptV2) =
            Alonzo.PlutusV2
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 11]
```diff
@@ -1585,6 +1609,7 @@ getScriptIntegrityHash = \case
         Cardano.ShelleyBasedEraAllegra -> Nothing
         Cardano.ShelleyBasedEraMary    -> Nothing
         Cardano.ShelleyBasedEraAlonzo  -> SafeHash.originalBytes <$> scriptIntegrityHashOfAlonzoTx tx
+        Cardano.ShelleyBasedEraBabbage -> SafeHash.originalBytes <$> scriptIntegrityHashOfBabbageTx tx
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs]:
```suggestion
        Cardano.ShelleyBasedEraBabbage ->
            SafeHash.originalBytes <$> scriptIntegrityHashOfBabbageTx tx
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 12]
```diff
@@ -1959,6 +1996,24 @@ toCardanoTxOut era = case era of
...
+            addrInEra
+            (Cardano.TxOutValue Cardano.MultiAssetInBabbageEra $ toCardanoValue tokens)
+            datumHash
+            refScript
+      where
+        refScript = Cardano.ReferenceScriptNone
+        datumHash = Cardano.TxOutDatumNone
+        addrInEra = tina "toCardanoTxOut: malformed address"
+            [ Cardano.AddressInEra (Cardano.ShelleyAddressInEra Cardano.ShelleyBasedEraBabbage)
+                <$> deserialiseFromRawBytes AsShelleyAddress addr
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs]:
```suggestion
            [ Cardano.AddressInEra
                (Cardano.ShelleyAddressInEra Cardano.ShelleyBasedEraBabbage)
                    <$> deserialiseFromRawBytes AsShelleyAddress addr
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs [hunk 13]
```diff
@@ -1012,11 +1014,11 @@ genConfig dir systemStart clusterEra logCfg = do
...
 
     -- we need to specify genesis file location every run in tmp
-    withAddedKey k v = withObject (pure . HM.insert k (toJSON v))
-    withHardForks era = withObject (pure . HM.union (HM.fromList hardForks))
+    withAddedKey k v = withObject (pure . Aeson.insert k (toJSON v))
+    withHardForks era = withObject (pure . Aeson.union (Aeson.fromList hardForks))
       where
         hardForks =
-            [ ("Test" <> T.pack (show hardFork) <> "AtEpoch", Yaml.Number 0)
+            [ (Aeson.fromText $ "Test" <> T.pack (show hardFork) <> "AtEpoch", Yaml.Number 0)
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs]:
```suggestion
            [
                ( Aeson.fromText $ "Test" <> T.pack (show hardFork) <> "AtEpoch"
                , Yaml.Number 0
                )
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 14]
```diff
@@ -596,6 +597,11 @@ newTransactionLayer networkId = TransactionLayer
                     signTransaction networkId acctResolver policyResolver
                     addressResolver inputResolver (body, wits)
                     & sealedTxFromCardano'
+                InAnyCardanoEra BabbageEra (Cardano.Tx body wits) ->
+                    signTransaction networkId acctResolver policyResolver
+                    addressResolver inputResolver (body, wits)
+                    & sealedTxFromCardano'
+
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:611]:
```suggestion
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 15]
```diff
@@ -2106,13 +2161,16 @@ unsealShelleyTx wtx = case W.cardanoTx wtx of
...
+    Cardano.InAnyCardanoEra BabbageEra tx ->
+        TxInMode tx BabbageEraInCardanoMode
 
 -- | Converts a 'ShelleyBasedEra' to the broader 'CardanoEra'.
 shelleyBasedToCardanoEra :: ShelleyBasedEra era -> CardanoEra era
 shelleyBasedToCardanoEra Cardano.ShelleyBasedEraShelley = ShelleyEra
 shelleyBasedToCardanoEra Cardano.ShelleyBasedEraAllegra = AllegraEra
 shelleyBasedToCardanoEra Cardano.ShelleyBasedEraMary    = MaryEra
 shelleyBasedToCardanoEra Cardano.ShelleyBasedEraAlonzo  = AlonzoEra
+shelleyBasedToCardanoEra Cardano.ShelleyBasedEraBabbage  = BabbageEra
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs]: Suggestion: use `LambdaCase` to remove the repetition of the function name:

```hs
shelleyBasedToCardanoEra :: ShelleyBasedEra era -> CardanoEra era
shelleyBasedToCardanoEra = \case
    Cardano.ShelleyBasedEraShelley -> ShelleyEra
    Cardano.ShelleyBasedEraAllegra -> AllegraEra
    Cardano.ShelleyBasedEraMary    -> MaryEra
    Cardano.ShelleyBasedEraAlonzo  -> AlonzoEra
    Cardano.ShelleyBasedEraBabbage -> BabbageEra
```

**Review by jonathanknowles — COMMENTED** [review]
Hi @Anviking 

Just leaving some suggestions (See comments.)

File: cabal.project [hunk 16]
```diff
@@ -222,25 +234,52 @@ source-repository-package
...
+  location: https://github.com/input-output-hk/hedgehog-extras
+  tag: 967d79533c21e33387d0227a5f6cc185203fe658
+  --sha256: 0rbqb7a64aya1qizlr3im06hdydg9zr6sl3i8bvqqlf7kpa647sd
+
+-- Needed for cardano-addresses
+source-repository-package
+     type: git
+     location: https://github.com/haskell-works/hw-aeson
+     tag: d99d2f3e39a287607418ae605b132a3deb2b753f
+     --sha256: 1vxqcwjg9q37wbwi27y9ba5163lzfz51f1swbi0rp681yg63zvn4
```
Comment (Anviking) [inline cabal.project]: Probably not needed now, will attempt removing.

File: cabal.project [hunk 17]
```diff
@@ -252,33 +291,49 @@ allow-newer:
...
 package cardano-config
-  flags: -systemd
+  flags: +systemd
 package cardano-node
-  flags: -systemd
+  flags: +systemd
+package plutus-core
+  ghc-options: -fexternal-interpreter
+  -- Needed to fix https://hydra.iohk.io/build/15694577/nixlog/2/tail
+  -- Cf. https://input-output-rnd.slack.com/archives/C21UF2WVC/p1624564618267800?thread_ts=1624563615.267000&cid=C21UF2WVC
```
Comment (Anviking) [inline cabal.project]: Will try reverting in case it wasn't necessary.

File: cabal.project [hunk 18]
```diff
@@ -252,33 +291,49 @@ allow-newer:
...
 
 -- ----------------------------------------------------------------
 -- Flags for dependencies
 
 package cardano-config
-  flags: -systemd
+  flags: +systemd
 package cardano-node
-  flags: -systemd
+  flags: +systemd
```
Comment (Anviking) [inline cabal.project:346]: 🧐 Not sure where this came from nor the implications of it. Do we want it?

edit: it's from https://github.com/input-output-hk/cardano-wallet/pull/3293

File: lib/core/cardano-wallet-core.cabal [hunk 19]
```diff
@@ -419,6 +421,9 @@ test-suite unit
     , wai
     , wai-extra
     , warp
+  if (flag(scrypt))
+    cpp-options: -DHAVE_SCRYPT
+    build-depends: scrypt
```
Comment (Anviking) [inline lib/core/cardano-wallet-core.cabal:426]: Ah, yes. The unit tests were still unconditionally dependent on `scrypt`, unlike the library, so this seems good. Haven't checked whether it passes, but this seems fair nonetheless.

File: lib/core/src/Cardano/Api/Gen.hs [hunk 20]
```diff
@@ -266,7 +269,26 @@ genTxInsCollateral era =
...
+genSlotNo32 :: Gen SlotNo32
+genSlotNo32 = do
+    offset <- genOffset
+    frequency
+        [ (20, pure $ SlotNo32 offset)
+        , (20, pure $ SlotNo32 (maxBound @Word32 - offset) )
+        , (60, SlotNo32 <$> arbitrary @Word32)
+        ]
+  where
+    genOffset = choose (0, 10_000)
```
Comment (Anviking) [inline lib/core/src/Cardano/Api/Gen.hs:291]: 👍

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 21]
```diff
@@ -1210,23 +1291,55 @@ _assignScriptRedeemers (toAlonzoPParams -> pparams) ti resolveInput redeemers tx
         case res of
             Left (UnknownTxIns ins) ->
                 Left $ ErrAssignRedeemersUnresolvedTxIns $ map fromShelleyTxIn (F.toList ins)
+            Left (BadTranslation translationError) ->
+                Left $ ErrAssignRedeemersTranslationError $ show translationError
```
Comment (Anviking) [inline lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs]: 👍

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 22]
```diff
@@ -1185,23 +1247,42 @@ _assignScriptRedeemers (toAlonzoPParams -> pparams) ti resolveInput redeemers tx
...
+        let costs = toCostModelsAsArray (Alonzo.unCostModels $ Alonzo._costmdls pparams')
         let systemStart = getSystemStart ti
 
-        epochInfo <- hoistEpochInfo (left ErrAssignRedeemersPastHorizon . runIdentity . runExceptT)
+        epochInfo <- hoistEpochInfo (left (T.pack . show . ErrAssignRedeemersPastHorizon) . runIdentity . runExceptT)
             <$> left ErrAssignRedeemersPastHorizon (toEpochInfo ti)
 
-        res <- evaluateTransactionExecutionUnits
-                pparams
+        let res = evaluateTransactionExecutionUnits
```
Comment (Anviking) [inline lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:1306]: 🧐 I think the `ErrAssignRedeemersPastHorizon` gets lost here. (`toEpochInfo` above won't throw it, which is a flaw of `TimeInterpreter`, but still, whether we get a PastHorizon or not would only be known when `evaluateTransactionExecutionUnits` actually tries to convert the validity interval.)

Will try to fix.

Comment (Anviking) [reply lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:1306]: Ah, it seems we no longer get the actual `PastHorizonException` from the ledger 😬 https://github.com/input-output-hk/cardano-ledger/blob/e290bf8d0ea272a51e9acd10adc96b4e12e00d37/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs#L106

Comment (Anviking) [reply lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:1306]: We probably should remove `ErrAssignRedeemersPastHorizon` then, and expose each constructor of `TranslationError` separately in the our API.

Pushed this for now: 3c62f15f3b7879b4e0460ab20647b0280e0a9d4c

File: cabal.project (hunk 18 above)
Comment (Anviking) [reply cabal.project:346]: I see the node repo has docs which says

> We recommend [systemd](https://www.freedesktop.org/wiki/Software/systemd/) on Linux

and has  `systemd = !pkgs.stdenv.hostPlatform.isMusl` in its nix config.

With this change we have `flags: +systemd` but which can be overridden to [false if isMusl](https://github.com/input-output-hk/cardano-wallet/blob/1f8742817b9c7a2ab4a819d3ed284931880c371d/nix/haskell.nix#L407-L409), which should be equivalent to the node's `= !isMusl`.

So while I don't grasp the implications, I'd guess this is fine.

**Review by paweljakubas — APPROVED** [review]
Let's merge this beast! ;-)

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs (hunk 22 above)
Comment (Anviking) [reply lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:1306]: Or perhaps we can minimise the api-breakage by still using the `past_horizon` error code but with less details than the full `ErrAssignRedeemersPastHorizon`. Will think/look into tomorrow.

File: lib/cli/test/unit/Cardano/CLISpec.hs [hunk 23]
```diff
@@ -54,7 +55,7 @@ import Data.Text
     ( Text )
 import Data.Text.Class
     ( FromText (..), TextDecodingError (..), toText )
-import Options.Applicative
+import "optparse-applicative" Options.Applicative
```
Comment (jonathanknowles) [inline lib/cli/test/unit/Cardano/CLISpec.hs:59]: This is the first time I've seen us use the `PackageImports` extension. It's very lucky that this extension exists! Though it's also frustrating that it has to exist.

File: flake.nix [hunk 24]
```diff
@@ -33,7 +33,7 @@
   ############################################################################
 
   inputs = {
-    nixpkgs.follows = "haskellNix/nixpkgs-2111";
+    nixpkgs.follows = "haskellNix/nixpkgs-unstable";
```
Comment (jonathanknowles) [inline flake.nix:36]: Do we know what the implications of following `nixpkgs-unstable` are?

For example, will it lead to more frequent fetching of dependencies? (i.e., when entering a `nix develop` shell, will we have to download dependencies more frequently than before?)

File: lib/core/src/Cardano/Api/Gen.hs [hunk 25]
```diff
@@ -1284,7 +1323,10 @@ genTxBodyContent era = do
         txBody = TxBodyContent
             { Api.txIns
             , Api.txInsCollateral = TxInsCollateralNone
+            , Api.txInsReference = TxInsReferenceNone
             , Api.txOuts
+            , Api.txTotalCollateral = TxTotalCollateralNone
+            , Api.txReturnCollateral = TxReturnCollateralNone
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Api/Gen.hs:1337]: @Anviking @paweljakubas 

We generate a value for `txIns` here. Do we need a task to generate values for `txInsCollateral`, `txTotalCollateral`, and `txReturnCollateral`, and `txInsReference`?

File: lib/core/src/Cardano/Wallet/Transaction.hs [hunk 26]
```diff
@@ -507,10 +506,10 @@ data ErrAssignRedeemers
     -- ^ The given redeemer target couldn't be located in the transaction.
     | ErrAssignRedeemersInvalidData Redeemer String
     -- ^ Redeemer's data isn't a valid Plutus' data.
-    | ErrAssignRedeemersPastHorizon PastHorizonException
-    -- ^ Evaluating the Plutus script failed past the visible horizon.
     | ErrAssignRedeemersUnresolvedTxIns [TxIn]
     -- ^ The transaction contains inputs which couldn't be resolved.
+    | ErrAssignRedeemersTranslationError String
+    -- ^ Mistranslating of hashes, credentials, certificates etc.
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Transaction.hs:514]: @Anviking Perhaps we can avoid this "stringly-typed" constructor by reusing `Cardano.Ledger.Alonzo.TxInfo.TranslationError` here?

We can defer the conversion to `String` to the `IsServerError` instance for `ErrAssignRedeemers`.

I've added 035fd118f60bb9b0b79b10fc7988347849649fce. Though feel free to remove if you feel it is inappropriate.

**Review by jonathanknowles — APPROVED** [review]
This looks reasonable to me.

I just have a few questions! (See comments.)

File: flake.nix (hunk 24 above)
Comment (Anviking) [reply flake.nix:36]: > Do we know what the implications of following nixpkgs-unstable are?

I don't entirely, but it seems to be needed, and I believe the consensus is that it's "fine".

>  (i.e., when entering a nix develop shell, will we have to download dependencies more frequently than before?)

No, the dependencies are locked in `flake.lock`, and we'd only need to download new dependencies when we bump... `haskell.nix` I think.

File: lib/cli/test/unit/Cardano/CLISpec.hs (hunk 23 above)
Comment (Anviking) [reply lib/cli/test/unit/Cardano/CLISpec.hs:59]: For the record, Sam elaborated further in: [604f06e](https://github.com/input-output-hk/cardano-wallet/pull/3280/commits/604f06e1fe02cd69bb119cdbc068237a4617ddc8)

File: lib/core/src/Cardano/Api/Gen.hs (hunk 25 above)
Comment (Anviking) [reply lib/core/src/Cardano/Api/Gen.hs:1337]: Hm:

- `txTotalCollateral` - might be added as part of ADP-1653
- `txReturnCollateral` - might be added as part of ADP-1653
- `txInsReference` - relevant to ADP-1655
- `txInsCollateral` - <s>I have added it to https://input-output.atlassian.net/browse/ADP-1374</s>

Comment (Anviking) [reply lib/core/src/Cardano/Api/Gen.hs:1337]: I don't think their addition should be _strictly_ required in ADP-1653 or ADP-1655 respectively, and at the same time I suspect creating a new ticket at the moment might just add noise.

Comment (Anviking) [reply lib/core/src/Cardano/Api/Gen.hs:1335]: 843a4070e82bbf96215e1bbc24e2c4ded68febb2

Comment (Anviking) [reply lib/core/src/Cardano/Api/Gen.hs:1337]: 0902c88eaefb7bc833134664a802e4a8ce49cc52

And actually we are already generating `txInsCollateral`, just at a later step (added a note).

File: lib/core/src/Cardano/Wallet/Transaction.hs (hunk 26 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Transaction.hs:514]: Thanks, that's better!

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs (hunk 22 above)
Comment (Anviking) [reply lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:1306]: I pushed 896c0680766b77fc761bdf0313d8ca25977bcbdb to minimise the API breakage from this PR and keep using the `past_horizon` error code. cc @jonathanknowles does it sound somewhat reasonable?

**Comment by Anviking** [comment]
I think it would be good to have #3328 or something similar before merging for the sake of removing `stack.yaml` and having an easy way of updating the `sha256` hashes — @sevanspowell objections, thoughts or do you see another way?

**Comment by sevanspowell** [comment]
> I think it would be good to have https://github.com/input-output-hk/cardano-wallet/pull/3328 or something similar before merging for the sake of removing stack.yaml and having an easy way of updating the sha256 hashes — @sevanspowell objections, thoughts or do you see another way?

No objection! Nice work :partying_face: I've merged #3328.

**Pull request merged**